module-container.md

```diff
@@ -0,0 +1,312 @@
+
+> Note: This document is highly Work-In-Progress. Please get involved. Comment discuss and add your own ideas.  
+
+## Definition
+
+The term "module container" builds upon "application container" coined by Docker. An application container focuses on running a single process in contrast to multiple processes per container. If the application consists of multiple processes they are spread out to different containers. A module container refines application container with the focus on being a good building block. In addition it suggests an even smaller granularity.
+
+## Related work
+
+In this field there is much prior and contemporary art:
+
+* [The 12 Factor App](http://12factor.net/) a widely cited site describing how to write good PaaS applications. Initiated by Heroku. Many of these principles apply for containers as well.
+
+* Cloud native: In the past year the term "cloud native applications" has gained popularity. A good introduction can be found in the free ebook [Migrating to Cloud Native Application Architectures](http://pivotal.io/platform/migrating-to-cloud-native-application-architectures-ebook)
+
+The folks from [Deis](http://deis.io/) have coined the term [cluster aware image](https://twitter.com/gabrtv/status/676218070357032961).
+
+## Properties of a module container
+
+There is no canonical definition of a "module container". Instead we gather a set of guiding properties and associated best practices:
+
+1. [Linux process](#1-linux-process)
+2. [Container API](#2-container-api)
+3. [Explicit interfaces](#3-explicit-interfaces)
+4. [Disposable](#4-disposable)
+5. [Immutable](#5-immutable)
+6. [Self-Contained](#6-self-contained)
+7. [Small](#small)
+
+### 1. Linux process
+
+Before we come up with to many new ideas we should acknowledge the fact that a container is foremost a Linux process. Therefore we should apply common standards and best practices for writing Unix tools which happen to be containers. These are _React to signals_, _Use standard streams_, _Handle arguments_:
+
+#### React to signals
+A container should react to the [signals](https://en.wikipedia.org/wiki/Unix_signal) which are sent to it. So our applications in the container should run in the foreground and catch signals and react appropriately. 
+
+Important signals to catch are:
+
+* `SIGINT`: E.g. send by Ctrl-C to interrupt / stop a running container.  
+* `SIGTERM`: Signals process termination e.g. send by `docker stop`. If necessary the container can do some final steps. After an usual grace period of 10 seconds the container is killed if it hasn't returned anymore
+
+**Example:**  
+Catching `SIGTERM` in Node.JS:  
+
+```javascript
+process.on('SIGTERM', function () {
+  console.log("Received SIGTERM. Exiting.")
+  server.close(function () {
+    process.exit(0);
+  });
+});
+``` 
+https://github.com/luebken/currentweather/blob/master/server.js#L49
+
+**Docker best practices:**
```

Comment:
I would actually not call this best practice but just something to be aware of. If you do it wrong but try to build a good container it may take a long time to find this.